internal/journal/volumegroupjournal.go

```diff
@@ -0,0 +1,429 @@
+/*
+Copyright 2024 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package journal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/google/uuid"
+)
+
+const (
+	defaultVolumeGroupNamingPrefix string = "csi-vol-group-"
+)
+
+type VolumeGroupJournal interface {
+	Connect(
+		monitors,
+		namespace string,
+		cr *util.Credentials) (*VolumeGroupJournalConfig, error)
+	Destroy()
+	setNamespace(ns string)
+	CheckReservation(
+		ctx context.Context,
+		journalPool,
+		reqName,
+		namePrefix string) (*VolumeGroupData, error)
+	UndoReservation(
+		ctx context.Context,
+		csiJournalPool,
+		snapshotGroupName,
+		reqName string) error
+	GetVolumeGroupAttributes(
+		ctx context.Context,
+		pool,
+		objectUUID string) (*VolumeGroupAttributes, error)
+	ReserveName(
+		ctx context.Context,
+		journalPool string,
+		journalPoolID int64,
+		reqName,
+		namePrefix string) (string, string, error)
+	AddVolumeSnapshotMapping(
+		ctx context.Context,
+		pool,
+		reservedUUID,
+		volumeID,
+		snapshotID string) error
+	RemoveVolumeSnapshotMapping(
+		ctx context.Context,
+		pool,
+		reservedUUID,
+		volumeID string) error
+}
+
+// VolumeGroupJournalConfig contains the configuration and connection details.
+type VolumeGroupJournalConfig struct {
```

Comment:
This should probably not be exported, only the functions of the interface should be available.